## Problem 12: Highly Divisible Triangular Numbers

The sequence of triangle numbers is generated by adding the natural numbers. So the 7th triangle number would be $1 + 2 + 3 + 4 + 5 + 6 + 7 = 28$. The first ten terms would be:

1, 3, 6, 10, 15, 21, 28, 36, 45, 55, ...

Let us list the factors of the first seven triangle numbers:

\begin{align}
 1&: 1\\
 3&: 1,3\\
 6&: 1,2,3,6\\
10&: 1,2,5,10\\
15&: 1,3,5,15\\
21&: 1,3,7,21\\
28&: 1,2,4,7,14,28\\
\end{align}

We can see that 28 is the first triangle number to have over five divisors.

What is the value of the first triangle number to have over five hundred divisors?

In [1]:
import numpy as np

In [2]:
''' 
Counts the number of divisors a particular number has. 
Only need to search up to the square root of the number.
Each factor we find counts for 2:
    1 for the factor itself
    1 for the number > sqrt(number) that is the other factor.
If the factor is the square root itself, then it only counts for one factor.

eg:
>>> number_of_divisors(12345)
>>> 8
'''

def number_of_divisors(number : int) -> int:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Counts the number of divisors a particular number has. 
        Only need to search up to the square root of the number.
        Each factor we find counts for 2:
            1 for the factor itself
            1 for the number > sqrt(number) that is the other factor.
        If the factor is the square root itself, then it only counts for one factor.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : number : The number to count the number of divisors for.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : int : n_divisors : The number of divisors.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> number_of_divisors(12345)
        >>> 8
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(number, int): raise ValueError('Please enter an integer > 0 for the number argument.')
    if number < 0: raise ValueError('Please enter an integer > 0 for the number argument.')
    
    n_divisors = 0
    for i in range(1, int(np.ceil(np.sqrt(number)))):
        if number% i == 0: n_divisors += 2
    if number % np.ceil(np.sqrt(number)) == 0: n_divisors -= 1
    
    return n_divisors

In [3]:
''' 
Generates triangle numbers and counts the number of divisors each one has.
If the number of divisors is over the target, it returns information about the solution.

eg:
>>> generate_triangle_numbers_find_divisors(50)
>>> (224, 25200, 90)
'''

def generate_triangle_numbers_find_divisors(target_divisors : int) -> list:
    '''
    --- Function Description --------------------------------------------------------------------------------------------------
        Generates triangle numbers and counts the number of divisors each one has.
        If the number of divisors is over the target, it returns information about the solution.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Inputs -------------------------------------------------------------------------------------------------------
        : int : target_divisors : The number of divisors to aim for.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Outputs ------------------------------------------------------------------------------------------------------
        : list : Returns the following information:
            : int : triangle_addition : The position of the triangle number.
            : int : triangle_number : The triangle number.
            : int : n_divisors : The number of divisors the triangle number has.
    ---------------------------------------------------------------------------------------------------------------------------
    
    --- Function Examples -----------------------------------------------------------------------------------------------------
        >>> generate_triangle_numbers_find_divisors(50)
        >>> (224, 25200, 90)
    ---------------------------------------------------------------------------------------------------------------------------
    '''
    
    # Check types of function inputs:
    if not isinstance(target_divisors, int): raise ValueError('Please enter an integer > 0 for the target_divisors argument.')
    if target_divisors < 0: raise ValueError('Please enter an integer > 0 for the target_divisors argument.')
    
    n_divisors = 1
    triangle_number = 1
    triangle_addition = 1
    
    while n_divisors < target_divisors:
        triangle_addition += 1
        triangle_number += triangle_addition
        n_divisors = number_of_divisors(triangle_number)
    
    return triangle_addition, triangle_number, n_divisors

In [4]:
problem_target_divisors = 500
solution = generate_triangle_numbers_find_divisors(problem_target_divisors)

In [5]:
print(f'The first triangle number to have more than {problem_target_divisors} divisors is {solution[1]:,}, which is the {solution[0]:,} triangle number.')
print(f'It has {solution[2]} divisors.')

The first triangle number to have more than 500 divisors is 76,576,500, which is the 12,375 triangle number.
It has 576 divisors.


### Problem 12 Solution:

The first triangle number to have more than 500 divisors is 76,576,500, which is the 12,375 triangle number. It has 576 divisors.